### DISCUSSION TITLE
The standard (BIF) library - developments since last Thursday

### DISCUSSION THREAD

**Discussion by rvjansen** [adesutherland/CREXX#136] [Progress Updates]
This is what happened in the meantime: the goal of prioritising the implementation of the handling of optional function arguments in cRexx was to enable us to split off work on the library to one part of the team while the architect works on other things. Most of it can be followed in [this thread](https://github.com/adesutherland/CREXX/discussions/133) 

It comes down to this:

- the linkage conventions were slightly changed to accommodate optional arguments, and to be precise, now the base register for the call contains the maximum number of arguments, so including arguments that are not specified
- the register type is used to indicate empty (non-specified) registers, for this there are new instructions like  brtpandt l33a,a4,1 to branch if there is a value. This works 'out of the box' with rexx source, and has to be changed in the current rxas bif implementations. An example is:

```
    /* here we need to substract the left-out optionals from the argcount */    
    brtpandt havethree,a3,1
    isub  a0,a0,1
havethree:
    brtpandt havefour,a4,1
    isub  a0,a0,1
havefour:   
    /* now we have the correct number */
```

This is how I fixed substr in the rxas version without too much change.

- I have renamed the lib/rxfns/zero directory to lib/rxfns/rexx because I am too old to remember that myself 
- There is a lib/rxfns/tests_functional directory that needs to have unified testcases for the rexx and rxas versions of a bif. They are, after all, meant to be called from rexx
- in the test.sh script there you can see how to run them, using the rexx script for each bif from tests_functional and 'linking' the .rxbin files from their respective subdirectories, rexx and rxas 
- Several existing rxas bifs need to be fixed still, I am doing that today
- Anyone can choose a bif to implement in rexx or rxas from the list at [bif-planning-list](https://github.com/adesutherland/CREXX/wiki/Rexx-levelb-bif-planning) , but please advertise the fact that you are working on it
- There was a quick fix from Adrian and some undue pressure from me (apologies!) to enable the automatic linking (well, in the sense that no hand-editing of the generated rxas was needed. The rxas test case (testbifs.rxas) was adapted but still segfaults due to the before mentioned needed change to the linkage convention

I think that is it more or less.

**Comment by rvjansen** [comment]
And, for completeness, we are working in f0035, and the needed compiler and instruction changes are merged in there, so you just need to rebuild your release- and debug versions from there.

**Comment by Peter-Jacob** [comment]
sorry for my ignorance, but I still don't get it. If I add a bif (rexx implementation) I place it in 
\...\lib\rxfns\rexx\ compile and assembly it there. Once it is completed, how do I test it? The test examples I found were in:
\...\lib\rxfns\tests_functional\. In which the rexx coding was like this:
```
...
...
 /* say substr("12345678",10,6,"é") */
/* if substr("12345678",10,6,"éé") \= 'éééééé' then say 'need exceptions for this' */

say "SUBSTR OK"

return
/* Substr() Procedure */
substr: procedure = .string
arg string1 = .string, start = .int, length = 0, pad = ' '

```
Meaning the bif function was duplicated for testing purposes from \...\lib\rxfns\rexx\. I am fine going this way (although not foolproof) , but maybe I misunderstand the development/testing process. Please enlighten me!

**Comment by rvjansen** [comment]
What I meant was that - we have one test function, for every bif. That one is in tests_functional; it is used to test the rexx version (if it is there) and the rxas version(s) if they are there. The test.sh script in tests_functional shows exactly how this works. So the function code lives in lib/rxfns/rexx and lib/rxfns/rxas; its test function lives in lib/rxfns/tests_functional - we test all implementations with one test function.

Now tests_performance will contain the performance tests (also one test, multiple target implementation) to help us choose which one goes to the actual library. That will be the fastest of the ones that have no functional errors.

best regards,

René.

**Comment by rvjansen** [comment]
and

> /* Substr() Procedure */
> substr: procedure = .string
> arg string1 = .string, start = .int, length = 0, pad = ' '

is the prototype of the implementations, its signature, which needs to go into the source - for the moment.

This works, now, for both the rexx implementation of substr() (3 errors) as the two rxas implementation (substr.rxas, 3 errorr, and oldsubstr.rxas, 1 error).

What Adrian changed last night, is that the rexx test automatically matches the advertised name of the function that is exposed. This is why it works now.

We can have a meeting tonight or tomorrow if you want me to show you.

René.

**Comment by Peter-Jacob** [comment]
A call tomorrow would be nice, I think it makes it easier than constantly guessing.

**Comment by rvjansen** [comment]
Is tomorrow 11.00 workable?

**Reply by Peter-Jacob** [reply]
it is, thanks for that!
